src/DomainGuest.sol

```diff
@@ -0,0 +1,347 @@
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    function file(bytes32 what, address data) external auth {
+        if (what == "end") end = EndLike(data);
+        else revert("DomainGuest/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function file(bytes32 what, uint256 data) external auth {
+        if (what == "dust") dust = data;
+        else revert("DomainGuest/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    // --- Canonical DAI Support ---
+
+    /// @notice Mint DAI and send to user
+    /// @param to The address to send the DAI to on the local domain
+    /// @param amount The amount of DAI to send [WAD]
+    function _deposit(address to, uint256 amount) internal {
+        vat.swell(address(this), _int256(amount * RAY));
+        daiJoin.exit(to, amount);
+
+        emit Deposit(to, amount);
+    }
+
+    /// @notice Withdraw DAI by burning local canonical DAI
+    /// @param to The address to send the DAI to on the remote domain
+    /// @param amount The amount of DAI to withdraw [WAD]
+    function _withdraw(address to, uint256 amount) internal {
+        require(dai.transferFrom(msg.sender, address(this), amount), "DomainGuest/transfer-failed");
+        daiJoin.join(address(this), amount);
+        vat.swell(address(this), -_int256(amount * RAY));
+
+        emit Withdraw(msg.sender, to, amount);
+    }
+
+    // --- MCD Support ---
+
+    /// @notice Record changes in Line
+    /// @param _lid Local ordering id
+    /// @param rad The new debt ceiling [RAD]
+    function _lift(uint256 _lid, uint256 rad) internal ordered(_lid) {
+        require(live == 1, "DomainGuest/not-live");
+
+        vat.file("Line", rad);
+
+        emit Lift(rad);
+    }
+
+    /// @notice Push surplus to the host dss
+    /// @dev Should be run by keeper on a regular schedule
+    function _surplus() internal returns (uint256 _rid, uint256 wad) {
+        require(live == 1, "DomainGuest/not-live");
+
+        _rid = rid++;
+
+        uint256 _dai = vat.dai(address(this));
+        uint256 _sin = vat.sin(address(this));
+        require(_dai > _sin, "DomainGuest/non-surplus");
+        unchecked { wad = (_dai - _sin) / RAY; } // Round against this contract for surplus
+        require(wad >= dust, "DomainGuest/dust");
+
+        // Burn the DAI and unload on the other side
+        vat.swell(address(this), -_int256(wad * RAY));
+
+        emit Surplus(wad);
+    }
+
+    /// @notice Push deficit to the host dss
+    /// @dev Should be run by keeper on a regular schedule
+    function _deficit() internal returns (uint256 _rid, uint256 wad) {
+        require(live == 1, "DomainGuest/not-live");
+
+        _rid = rid++;
+
+        uint256 _dai   = vat.dai(address(this));
+        uint256 _sin   = vat.sin(address(this));
+        uint256 _dsin  = dsin;
```

Comment:
`_dsin` does not seem to be necessary.

Reply:
It's used twice, is for avoiding reading twice the storage, even if there is warm up it is more gas expensive.